app/services/course/assessment/submission/auto_grading_service.rb

```diff
@@ -38,6 +50,13 @@ def grade_answers(submission)
   def grade_answer(answer)
     raise ArgumentError if answer.changed?
     answer.auto_grade!
+    # Catch errors if answer is in attempting state, caused by a race condition where
+    # a new attempting answer is created while the submission is finalised, but before the
+    # autograding job is executed.
+  rescue IllegalStateError
```

Comment:
> For consistency, would it be better to allow attempting answers to be evaluated too? 

Would advise against this, as there are use cases where the `IllegalStateError` is valid. From a business logic sense it would be better for us to disallow this by default in the code, and specially `rescue` it if need be, so that new developers won't accidentally call `#auto_grade!` on `attempting` answers.

> I think it'll be clearer to explicitly allow it in the Answer model's state transition definition. That'll also keep the state transition logic in 1 place and prevent unexpected surprises. (The next person might use this elsewhere and wonder why grading `attempting` answers doesn't throw an exception like the model definition suggests.)

If the intent is on better readability and trying to figure things out, I disagree also because this fix was tailored to resolve the race condition introduced in this AutoGradingService. In case of future refactoring/changes, I would expect that the developer does a `git blame`, reads the commit message, reads the PR, and reads the issue before deciding on what to modify. I've specially added a lot more information/context into these areas as well. 


Any thoughts @allenwq ?